src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+      return false;
+    return true;
+  }
+  void increase_tokens(qos_entry& it, const ceph::timespan curr_timestamp,
+                       const RGWQoSInfo* info)
+  {
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
+    if(it.first_run)
+    {
+      it.write.ops = info->max_write_ops * fixed_float;
+      it.write.bw = info->max_write_bytes * fixed_float;
+      it.read.ops = info->max_read_ops * fixed_float;
+      it.read.bw = info->max_read_bytes * fixed_float;
+      it.ts = curr_timestamp;
+      it.first_run = false;
+      return;
+    }
+    else if(curr_timestamp > it.ts && minimum_time_reached(curr_timestamp, it))
+    {
+      int64_t time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(curr_timestamp - it.ts).count() / 60.0 / std::milli::den * fixed_float; // / 60 to make it work with 1 min token bucket
+      it.ts = curr_timestamp;
+      int64_t write_ops = info->max_write_ops * time_in_ms;
+      int64_t write_bw = info->max_write_bytes * time_in_ms;
+      int64_t read_ops = info->max_read_ops * time_in_ms;
+      int64_t read_bw = info->max_read_bytes * time_in_ms;
+      it.read.ops = std::min(info->max_read_ops * fixed_float, read_ops + it.read.ops);
+      it.read.bw = std::min(info->max_read_bytes * fixed_float, read_bw + it.read.bw);
+      it.write.ops = std::min(info->max_write_ops * fixed_float, write_ops + it.write.ops);
+      it.write.bw = std::min(info->max_write_bytes * fixed_float, write_bw + it.write.bw);
+    }
+  }
+
+  bool should_rate_limit_write(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it.write_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it.write_bytes() < 0 && bw_limit > 0))
+    {
+      return true;
+    }
+
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it.write.ops -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+
+  public:
+    RateLimiter(const RateLimiter&) = delete;
+    RateLimiter& operator =(const RateLimiter&) = delete;
+    RateLimiter(RateLimiter&&) = delete;
+    RateLimiter& operator =(RateLimiter&&) = delete;
+    RateLimiter() = delete;
+    RateLimiter(std::atomic_bool *replacing, std::condition_variable *cv)
+      : replacing(replacing), cv(cv)
+    {
+      // prevents rehash, so no iterators invalidation
+      qos_entries.max_load_factor(1000);
+    };
+
+    bool should_rate_limit(const char *method, const std::string& key, ceph::coarse_real_time curr_timestamp, const RGWQoSInfo* qos_info) {
+      if(key.empty() || key.length() == 1 || !qos_info->enabled)
+        return false;
+      bool is_read = is_read_op(method);
+      // it is possible that some requests will have the same ts, because we are using coarse real time
+      auto& it = find_or_create(key);
+      auto curr_ts = curr_timestamp.time_since_epoch();
+      std::unique_lock lock(it.ts_lock);
```

Comment:
maybe better to move the record lock inside the functions, and use it only when entry values are modified?

Reply:
we will need RWlocks anyway, I don't see how it helps, values modification here is quite common